### PR TITLE
Add Transaction Fee RPC to Statemint/Statemine

### PR DESCRIPTION
**Pull request by hamidra**
changes to add the support for payment rpc to parachains to address #542

### CODE REVIEW

File: polkadot-parachains/Cargo.toml [hunk 1]
```diff
@@ -64,6 +64,9 @@ substrate-prometheus-endpoint = { git = "https://github.com/paritytech/substrate
 
 # RPC related dependencies
 jsonrpc-core = "15.1.0"
+txpool-api = { package = "sc-transaction-pool-api", git = "https://github.com/paritytech/substrate", branch = "master" }
```
Comment (bkchr) [inline polkadot-parachains/Cargo.toml:67]:
```suggestion
sc-transaction-pool-api = { git = "https://github.com/paritytech/substrate", branch = "master" }
```

File: polkadot-parachains/src/rpc.rs [hunk 2]
```diff
@@ -0,0 +1,79 @@
+// Copyright 2019-2021 Parity Technologies (UK) Ltd.
```
Comment (bkchr) [inline polkadot-parachains/src/rpc.rs:1]:
```suggestion
// Copyright 2021 Parity Technologies (UK) Ltd.
```

File: polkadot-parachains/src/service.rs [hunk 3]
```diff
@@ -51,12 +53,8 @@ use sp_runtime::{
 use std::sync::Arc;
 use substrate_prometheus_endpoint::Registry;
 
-pub use sc_executor::NativeExecutor;
-
 type BlockNumber = u32;
 type Header = sp_runtime::generic::Header<BlockNumber, sp_runtime::traits::BlakeTwo256>;
-pub type Block = sp_runtime::generic::Block<Header, sp_runtime::OpaqueExtrinsic>;
```
Comment (bkchr) [inline polkadot-parachains/src/service.rs:58]: Why?

File: polkadot-parachains/src/service.rs [hunk 4]
```diff
@@ -28,7 +27,9 @@ use cumulus_primitives_core::{
 	relay_chain::v1::{Hash as PHash, PersistedValidationData},
 	ParaId,
 };
+pub use polkadot_primitives::v1::{AccountId, Balance, Block, Hash, Nonce};
```
Comment (bkchr) [inline polkadot-parachains/src/service.rs:30]: This is not correct! 

We want to use the parachain specific types and not the types from Polkadot.

File: polkadot-parachains/src/service.rs [hunk 5]
```diff
@@ -51,12 +53,8 @@ use sp_runtime::{
 use std::sync::Arc;
 use substrate_prometheus_endpoint::Registry;
 
-pub use sc_executor::NativeExecutor;
```
Comment (bkchr) [inline polkadot-parachains/src/service.rs:54]: Why did you moved this?

File: polkadot-parachains/src/service.rs (hunk 4 above)
Comment (hamidra) [reply polkadot-parachains/src/service.rs]: Makes sense! We probably want to factor them out as a module like to make them compatible. for now defined them as separate types.

File: polkadot-parachains/src/service.rs (hunk 5 above)
Comment (hamidra) [reply polkadot-parachains/src/service.rs:54]: I guess it is analyzer sorting the imports.

File: polkadot-parachains/src/rpc.rs [hunk 6]
```diff
@@ -20,14 +20,24 @@
...
+/// Alias to 512-bit hash when used in the context of a transaction signature on the chain.
+pub type Signature = sp_runtime::MultiSignature;
+/// Some way of identifying an account on the chain. We intentionally make it equivalent
+/// to the public key of our transaction signing scheme.
+pub type AccountId = <<Signature as sp_runtime::traits::Verify>::Signer as sp_runtime::traits::IdentifyAccount>::AccountId;
+type BlockNumber = u32;
+type Header = sp_runtime::generic::Header<BlockNumber, sp_runtime::traits::BlakeTwo256>;
+pub type Block = sp_runtime::generic::Block<Header, sp_runtime::OpaqueExtrinsic>;
+type Balance = u128;
+type Nonce = u32;
```
Comment (shawntabrizi) [inline polkadot-parachains/src/rpc.rs:39]: i dont think this is right either. shouldnt be that we are redefining all these types here.

should make the function generic over these types, or properly import them

File: polkadot-parachains/src/rpc.rs [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
```
Comment (bkchr) [inline polkadot-parachains/src/rpc.rs:39]: You could make them public in service.rs and import them from there.

File: polkadot-parachains/src/service.rs [hunk 8]
```diff
@@ -29,6 +28,7 @@ use cumulus_primitives_core::{
 	ParaId,
 };
 
+use crate::rpc as parachain_rpc;
```
Comment (bkchr) [inline polkadot-parachains/src/service.rs:31]: Why the rename? Either name the file directly this way or remove this please.

File: polkadot-parachains/src/rpc.rs (hunk 7 above)
Comment (hamidra) [reply polkadot-parachains/src/rpc.rs:39]: doesn't this create cyclic dependency?! does rust support cyclic dependency?!

File: polkadot-parachains/src/rpc.rs (hunk 6 above)
Comment (hamidra) [reply polkadot-parachains/src/rpc.rs]: Moved them to primitives module and imported the common primitives in rpc.rs and service.rs.

File: polkadot-parachains/src/rpc.rs (hunk 7 above)
Comment (hamidra) [reply polkadot-parachains/src/rpc.rs]: moved and made them public in a new primitives module and imported the common primitives in rpc.rs and service.rs.

**Comment by Swader** [comment]
Is there any way we can move this along please? What can we do to help? This is a huge blocker for us.

File: polkadot-parachains/src/primitives.rs [hunk 9]
```diff
@@ -0,0 +1,11 @@
...
+pub type Signature = sp_runtime::MultiSignature;
+/// Some way of identifying an account on the chain. We intentionally make it equivalent
+/// to the public key of our transaction signing scheme.
+pub type AccountId = <<Signature as sp_runtime::traits::Verify>::Signer as sp_runtime::traits::IdentifyAccount>::AccountId;
+pub type BlockNumber = u32;
+pub type Header = sp_runtime::generic::Header<BlockNumber, sp_runtime::traits::BlakeTwo256>;
+pub type Block = sp_runtime::generic::Block<Header, sp_runtime::OpaqueExtrinsic>;
+pub type Hash = sp_core::H256;
+pub type Balance = u128;
+pub type Nonce = u32;
```
Comment (shawntabrizi) [inline polkadot-parachains/src/primitives.rs:11]: I made https://github.com/paritytech/cumulus/pull/577 as an improvement to this file.

Comment (shawntabrizi) [reply polkadot-parachains/src/primitives.rs:11]: Also in the future, need to add the copyright headers when adding new files, fyi

**Review by shawntabrizi — APPROVED** [review]

**Comment by hamidra** [comment]
bot merge

**Comment by hamidra** [comment]
bot merge

**Comment by Swader** [comment]
Hi, is there a specific reason why this wasn't included in the V3 release?

**Comment by shawntabrizi** [comment]
@Swader it was just merged, and release process was already in the pipeline. It will be included in the next release.

**Comment by Swader** [comment]
Sure, I will test our distribution script locally and report back

**Comment by DikaCream** [comment]
Hi, is there a specific date the next release of statemint including this request ?

**Comment by hamidra** [comment]
> Hi, is there a specific date the next release of statemint including this request ?

This will be in the next Statemine/mint release which will happen after the polkadot release. Next polkadot release is about to start, so hopefully soon after that this should be available on statemint.

**Pull request merged**